composeApp/src/main/kotlin/io/github/couchtracker/tmdb/BatcheableTmdbDownload.kt

```diff
@@ -0,0 +1,181 @@
+import kotlin.time.Duration
+import kotlin.time.Instant
+
+class BatchableDownloader<T : Any, Req, Res>(
+    val logTag: String,
+    val loadFromCache: (cache: TmdbCache) -> Query<TmdbTimestampedEntry<T>>,
+    val putInCache: (cache: TmdbCache, TmdbTimestampedEntry<T>) -> Unit,
+    val prepareRequest: (Req) -> Req,
+    val extractFromResponse: (Res) -> ApiResult<T>,
+    val expiration: Duration = TMDB_CACHE_EXPIRATION_DEFAULT,
+    val prefetch: Duration = expiration * TMDB_CACHE_PREFETCH_THRESHOLD,
+)
+
+typealias LocalizedQueryBuilder<ID, T> = (ID, TmdbLanguage, (T, Instant) -> TmdbTimestampedEntry<T>) -> Query<TmdbTimestampedEntry<T>>
+typealias QueryBuilder<ID, T> = (ID, (details: T, lastUpdate: Instant) -> TmdbTimestampedEntry<T>) -> Query<TmdbTimestampedEntry<T>>
+
+class BatchableDownloaderBuilder<ID, Req, Res>(
+    val id: ID,
+    val logTag: String,
+    val prepareRequest: (Req) -> Req,
+    val expiration: Duration = TMDB_CACHE_EXPIRATION_DEFAULT,
+    val prefetch: Duration = expiration * TMDB_CACHE_PREFETCH_THRESHOLD,
+) {
+
+    inner class Step2<T : Any>(
+        val extractFromResponse: (Res) -> ApiResult<T>,
+    ) {
+        fun localized(
+            language: TmdbLanguage,
+            loadFromCacheFn: (TmdbCache) -> LocalizedQueryBuilder<ID, T>,
+            putInCacheFn: (TmdbCache) -> (ID, TmdbLanguage, T, Instant) -> QueryResult<Long>,
+        ) = BatchableDownloader(
+            logTag = "$language-$logTag",
+            loadFromCache = { cache -> loadFromCacheFn(cache)(id, language, ::TmdbTimestampedEntry) },
+            putInCache = { cache, data -> putInCacheFn(cache)(id, language, data.value, data.lastUpdate) },
+            prepareRequest = prepareRequest,
+            extractFromResponse = extractFromResponse,
+            expiration = expiration,
+            prefetch = prefetch,
+        )
+
+        fun notLocalized(
+            loadFromCacheFn: (TmdbCache) -> QueryBuilder<ID, T>,
+            putInCacheFn: (TmdbCache) -> (ID, T, Instant) -> QueryResult<Long>,
+        ) = BatchableDownloader(
+            logTag = logTag,
+            loadFromCache = { cache -> loadFromCacheFn(cache)(id, ::TmdbTimestampedEntry) },
+            putInCache = { cache, data -> putInCacheFn(cache)(id, data.value, data.lastUpdate) },
+            prepareRequest = prepareRequest,
+            extractFromResponse = extractFromResponse,
+            expiration = expiration,
+            prefetch = prefetch,
+        )
+    }
+
+    fun <T : Any> extractResultFromResponse(extractFromResponse: (Res) -> ApiResult<T>): Step2<T> {
+        return Step2(extractFromResponse)
+    }
+
+    fun <T : Any> extractFromResponse(extractFromResponse: (Res) -> T): Step2<T> {
+        return extractResultFromResponse { response ->
+            Result.Value(extractFromResponse(response))
+        }
+    }
+
+    fun <T : Any> extractNonNullFromResponse(fieldName: String, extractFromResponse: (Res) -> T?): Step2<T> {
```

Comment:
If you want an additional optimization to avoid having to pass the field name:
```kotlin
    fun <T : Any> extractNonNullFromResponse(property: KProperty1<Res, T?>): Step2<T> {
        return extractNonNullFromResponse(fieldName = property.name, extractFromResponse = property::invoke)
    }
```

Callable with, e.g.:
```kotlin
.extractNonNullFromResponse(TmdbMovieDetail::credits)
```